### ISSUE TITLE
[A11Y] Validation error message are not read politely

### ISSUE THREAD

**Issue by jamesshannon** [ampproject/amphtml#12717]
Another A11Y issue from a partner.

They currently have a form validation javascript library which, on validation error, adds `aria-describedby` to the input error field on errors. This causes the screen reader to describe the input element and also read the error message. For example, given HTML similar to:

```
<input type="tel" id="zipCode" required pattern="\d{5}([-\s]?\d{4})?">
    <p visible-when-invalid="valueMissing" validation-for="zipCode">Enter a valid 5-digit or 9-digit ZIP Code.</p>
    <p visible-when-invalid="patternMismatch" validation-for="zipCode">Enter a valid 5-digit or 9-digit ZIP Code.</p>
```

Then iOS Safari VO will read out:

> ZIP Code  edit  invalid entry  required  error Enter a valid 5-digit or 9-digit ZIP Code.

Because the focus returns back to the input, which now has `aria-describedby`.

AMP doesn't add `aria-describedby` and thus VO doesn't read out the error message by default. 

One suggestion is to add `role=alert` to the error message, but then that causes VO to read the error message and the context (ie, the text field). It begins to read the field but then interrupts that with the alert.

`aria-live="polite" role="region"` should cause the browser to read both, and this is true on Mac Safari VO, but this is not the case on iOS Safari VO -- it behaves like `role=alert`.

Thus, I see two solutions I'd like to propose, and a workaround:

1. Add `aria-describedby` to the element which fails validation. This would require determining the element which matches the appropriate `visible-when-invalid` and `validation-for` attributes.
2. Add field-level validation triggers. And an action that allows attribute setting. Thus the element would look like this:

```
<input type="tel" id="zipCode" required pattern="\d{5}([-\s]?\d{4})?"
  trigger="
       validation-fail-empty:addAriaAttribute('describedby', 'zipCode-errormsg-empty'),
       validation-fail-pattern:addAriaAttribute('describedby', 'zipcode-errormsg-pattern'),
       validation-succeed:removeAriaAttribute('describedby')
   "
>
```

The workaround which I've proposed is to change the error text to include the description of the field:

```
<p visible-when-invalid="valueMissing" validation-for="zipCode">
****Zip Code field is required and invalid.***** Enter a valid 5-digit or 9-digit ZIP Code.
</p>
```

**Comment by ampprojectbot** [comment]
This issue hasn't been  updated in awhile. @aghassemi Do you have any updates?

**Comment by aghassemi** [comment]
@cvializ I'm upping the priority of this. We should ensure custom validation errors behave the same as native browser ones for accessibility.

**Comment by wassgha** [comment]
@cvializ I am currently working on this issue and I took the first route of adding `describedby` to the invalid element, however I ran into a problem related to the fact that `describedby` is expecting element IDs where as the error description elements might not necessarily have an ID. Is there a way to reinforce an ID on all error description elements or should we generate unique IDs on the fly? Open to suggestions.

**Comment by cvializ** [comment]
We should generate IDs. I think somewhere there might be a generic method to generate unique IDs for a11y, but if not we should create one. I found a specific one in `media-pool.js` but a general one would follow a similar strategy.

https://github.com/ampproject/amphtml/blob/bb5a459521d87a2677ade71a02d969035cd625ac/extensions/amp-story/0.1/media-pool.js#L264-L270
